### ISSUE TITLE
Ability to pull OAI-PMH metadata (and view records)

### ISSUE THREAD

**Issue by mlhale7** [notch8/utk-hyku#664]
# Story
ref. #665 
I am unable to pull metadata using OAI-PMH with my regular tools (https://github.com/vphill/pyoaiharvester) and I do not see individual records when navigating the feed (https://digitalcollections.lib.utk.edu/catalog/oai) in the browser. There appears to be an identifier issue that is causing no records to be retrievable. 

# Acceptance Criteria

- [x] Ability to pull OAI records in oai_dc or mods using standard command line tools or Repox
- [x] Ability to see individual records in the browser by navigating https://digitalcollections.lib.utk.edu/catalog/oai 

# Screenshots / Video

When I use pyoaiharvester, I get a ZeroDivisionError. Here's a screenshot showing the command and error:

![commandforHykuOAIwithErrors](https://github.com/user-attachments/assets/cb9846af-1242-4f57-8b44-2cfae583cbb5)

When I click on "oai_dc" in the browser to retrieve an individual record, I get the error "idDoesNotExist." Here's a screenshot:

![Screenshot 2024-07-23 at 11 45 25 AM](https://github.com/user-attachments/assets/d5574d87-7bde-4d83-9fc1-9c774223e834)

Finally, looking in the browser at https://digitalcollections.lib.utk.edu/catalog/oai, I am seeing records for attachments that I would not expect (PRESERVE, MODS, etc.) We just want a single record to appear for each digital asset.

![Screenshot 2024-07-23 at 12 51 54 PM](https://github.com/user-attachments/assets/ecb53b32-fdb6-45f0-8fc9-8f1f40a13df4)

# Testing Instructions and Sample Files

-

# Notes

A conjecture - potentially this issue was introduced when we changed the URL to "digitalcollections.lib.utk.edu"?

**Comment by mlhale7** [comment]
According to Repox (which we use for DPLA harvesting), our URL does not exist (I've tried both http and https just in case).

![Screenshot 2024-07-23 at 1 05 20 PM](https://github.com/user-attachments/assets/3070dc66-6bc0-48b1-9d4e-824e5e722424)

**Comment by kirkkwang** [comment]
An update:

Using the harvest (https://github.com/vphill/pyoaiharvester) as it was not working because https://digitalcollections.lib.utk.edu/ still has the HTTP Auth (where you need to enter the username and password before you can see the site).  Modifying the harvester's code to allow you to pass it in via the URL like http://username:password@example.com/ will get past that issue.

However, we are quickly met with a different issue.  There seems to be an issue with the SSL certificates.

![Image](https://github.com/user-attachments/assets/c9b520a8-e075-4228-ad50-157c6cb8f0d1)

Details in this [report](https://github.com/user-attachments/files/16460551/SSL.Server.Test_.digitalcollections.lib.utk.edu.Powered.by.Qualys.SS_.-.www.ssllabs.com.pdf).

Rob said he will address the cert issue.  After that is fixed we can try the harvester tool again.

**Comment by kirkkwang** [comment]
To summarize the findings here:

There are two things at play here that makes the `pyoaiharvester` not work with the production site.

1. Currently on the the `digitalcollections` tenant has basic auth turned on (the pop up where you are prompted for the username/password).  The `pyoaiharvester` tool does not allow passing in basic auth through the URL.  The assumption is that this will be turned off for launch so this should not be an issue.
2. We use [CrowdSec](https://www.crowdsec.net/about) as part of our security measures.  CrowdSec seems to be blocking the User-Agent pyoaiharvester/3.0.  This may be because it has been reported as potentially malicious or exhibiting suspicious behavior.  A workaround is to change the User-Agent in the script to something like `utklibraryoai`.  We are investigating the specific reasons for this block and assessing any necessary adjustments to our security configuration.

**Comment by mlhale7** [comment]
Thanks to @kirkkwang, I was able to successfully pull records in oai_dc and mods format. I'll comment back once I inspect these files more.

**Comment by mlhale7** [comment]
This may need to be a separate ticket, but I'm finding some odd records in the OAI. For instance:
```
    <record>
        <header>
            <identifier>oai:hyku:9a77b15d-554d-4dfc-a49f-09fbcee8118c</identifier>
            <datestamp>2024-07-22T23:24:01Z</datestamp>
            <setSpec>collection:admin_set/default</setSpec>
        </header>
        <metadata>
            <oai_dc:dc
                xsi:schemaLocation="http://www.openarchives.org/OAI/2.0/oai_dc/ http://www.openarchives.org/OAI/2.0/oai_dc.xsd"
                xmlns:oai_dc="http://www.openarchives.org/OAI/2.0/oai_dc/"
                xmlns:dc="http://purl.org/dc/elements/1.1/"
                xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
                <dc:description>OCR for alumnus:1507298774</dc:description>
                <dc:title>OCR</dc:title>
            </oai_dc:dc>
        </metadata>
    </record>
```

AND

```
    <record>
        <header>
            <identifier>oai:hyku:2f4595f1-7611-490a-b4b6-94336055d037</identifier>
            <datestamp>2024-07-22T23:23:58Z</datestamp>
            <setSpec>collection:admin_set/default</setSpec>
        </header>
        <metadata>
            <oai_dc:dc
                xsi:schemaLocation="http://www.openarchives.org/OAI/2.0/oai_dc/ http://www.openarchives.org/OAI/2.0/oai_dc.xsd"
                xmlns:oai_dc="http://www.openarchives.org/OAI/2.0/oai_dc/"
                xmlns:dc="http://purl.org/dc/elements/1.1/"
                xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
                <dc:description>TRANSCRIPT for jsevier:9</dc:description>
                <dc:title>TRANSCRIPT</dc:title>
            </oai_dc:dc>
        </metadata>
    </record>
    <record>
        <header>
            <identifier>oai:hyku:f47c15e5-5055-44b7-9b55-526b3e3bfc68</identifier>
            <datestamp>2024-07-22T23:23:58Z</datestamp>
            <setSpec>collection:admin_set/default</setSpec>
        </header>
        <metadata>
            <oai_dc:dc
                xsi:schemaLocation="http://www.openarchives.org/OAI/2.0/oai_dc/ http://www.openarchives.org/OAI/2.0/oai_dc.xsd"
                xmlns:oai_dc="http://www.openarchives.org/OAI/2.0/oai_dc/"
                xmlns:dc="http://purl.org/dc/elements/1.1/"
                xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
                <dc:description>TEI for jsevier:9</dc:description>
                <dc:title>TEI</dc:title>
            </oai_dc:dc>
        </metadata>
    </record>
    <record>
        <header>
            <identifier>oai:hyku:1211e8b1-040f-439d-b5d0-f1ad54f42e8d</identifier>
            <datestamp>2024-07-22T23:23:59Z</datestamp>
            <setSpec>collection:admin_set/default</setSpec>
        </header>
        <metadata>
            <oai_dc:dc
                xsi:schemaLocation="http://www.openarchives.org/OAI/2.0/oai_dc/ http://www.openarchives.org/OAI/2.0/oai_dc.xsd"
                xmlns:oai_dc="http://www.openarchives.org/OAI/2.0/oai_dc/"
                xmlns:dc="http://purl.org/dc/elements/1.1/"
                xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
                <dc:description>OBJ for jsevier:25</dc:description>
                <dc:title>OBJ</dc:title>
            </oai_dc:dc>
        </metadata>
    </record>
    <record>
        <header>
            <identifier>oai:hyku:72f7889f-da05-4363-a42a-c74354351672</identifier>
            <datestamp>2024-07-22T23:24:00Z</datestamp>
            <setSpec>collection:admin_set/default</setSpec>
        </header>
        <metadata>
            <oai_dc:dc
                xsi:schemaLocation="http://www.openarchives.org/OAI/2.0/oai_dc/ http://www.openarchives.org/OAI/2.0/oai_dc.xsd"
                xmlns:oai_dc="http://www.openarchives.org/OAI/2.0/oai_dc/"
                xmlns:dc="http://purl.org/dc/elements/1.1/"
                xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
                <dc:description>OBJ for jsevier:2</dc:description>
                <dc:title>OBJ</dc:title>
            </oai_dc:dc>
        </metadata>
    </record>
```

I'm including @laritakr here for informational purposes. None of these types of resources (OBJ, Transcript, OCR, etc) should be present in OAI. We just want the main record. Getting rid of all of these extra records would also make pulling OAI a lot faster. Ultimately the issue is that I would need to find a way to exclude all of these extra attachments for DPLA ingests etc. if not removed as this information is not needed.

**Comment by mlhale7** [comment]
Here's another odd record:

```
    <record>
        <header>
            <identifier>oai:hyku:24ac22aa-106b-4a88-a346-9e264d13d972</identifier>
            <datestamp>2023-09-01T03:59:26Z</datestamp>
            <setSpec>collection:admin_set/default</setSpec>
        </header>
        <metadata>
            <oai_dc:dc
                xsi:schemaLocation="http://www.openarchives.org/OAI/2.0/oai_dc/ http://www.openarchives.org/OAI/2.0/oai_dc.xsd"
                xmlns:oai_dc="http://www.openarchives.org/OAI/2.0/oai_dc/"
                xmlns:dc="http://purl.org/dc/elements/1.1/"
                xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
                <dc:publisher>utk</dc:publisher>
                <dc:rights>http://rightsstatements.org/vocab/InC/1.0/</dc:rights>
                <dc:title>504 error (shana)</dc:title>
            </oai_dc:dc>
        </metadata>
    </record>
```

**Comment by mlhale7** [comment]
HOCR also is not something we want a record for:

```
<record><header><identifier>oai:hyku:ccb61e6c-0c22-471c-9a22-e4dfe2953c62</identifier><datestamp>2024-07-24T05:13:52Z</datestamp><setSpec>collection:admin_set/default</setSpec></header><metadata><mods version="3.5" xsi:schemaLocation="http://www.loc.gov/mods/v3 http://www.loc.gov/standards/mods/v3/mods-3-5.xsd" xmlns="http://www.loc.gov/mods/v3" xmlns:xlink="http://www.w3.org/1999/xlink" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
  <titleInfo>
    <title>HOCR</title>
  </titleInfo>
  <identifier type="uuid">ccb61e6c-0c22-471c-9a22-e4dfe2953c62</identifier>
  <originInfo/>
  <physicalDescription/>
  <subject/>
  <subject>
    <cartographics/>
  </subject>
  <location>
    <url usage="primary" access="object in context">https://digitalcollections.lib.utk.edu/concern/attachments/ccb61e6c-0c22-471c-9a22-e4dfe2953c62</url>
    <url access="preview" xlink:href="https://digitalcollections.lib.utk.edu/assets/work-ff055336041c3f7d310ad69109eda4a887b16ec501f35afc0a547c4adb97ee72.png"/>
  </location>
  <recordInfo>
    <recordIdentifier>ccb61e6c-0c22-471c-9a22-e4dfe2953c62</recordIdentifier>
    <recordOrigin>https://digitalcollections.lib.utk.edu/catalog/oai</recordOrigin>
    <recordCreationDate>2024-05-17T19:45:38Z</recordCreationDate>
    <recordChangeDate>2024-05-17T21:53:27Z</recordChangeDate>
  </recordInfo>
</mods></metadata></record>
```

**Comment by laritakr** [comment]
Restricting the types of works that show in your OAI feed will should be a new ticket, as it is separate from the requirements of this ticket. 

This is due to the way the child works are created to allow additional metadata for file sets in your repo. We will need to identify which specific information we need to exclude and override standard OAI behavior.

**Comment by mlhale7** [comment]
@kirkkwang - I was able to pull both MODS and DC. Given that all of the sets have to be pulled each time, right now the time needed to get OAI is a bit restrictive, but this will be addressed when the ability to pull separate collections is added (#680). I approve the work completed in this ticket.

**Comment by kirkkwang** [comment]
Great! thank you @mlhale7

**Comment by josh-morgan117** [comment]
@mlhale7 and @kirkkwang I'm getting a "We're sorry, something went wrong" when trying to pull the mods for a record. 

eg https://digitalcollections.lib.utk.edu/catalog/oai?verb=GetRecord&metadataPrefix=mods&identifier=oai:hyku:950515c2-252b-4282-b6ff-132b359567b4

**Comment by mlhale7** [comment]
@kirkkwang @josh-morgan117 I'm not able to pull OAI-PMH using the same process I was before. I'm getting the error:
```
ssl.SSLCertVerificationError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1122)
```
Is it possible that the cert is having another issue?

**Comment by orangewolf** [comment]
@mlhale7 this should be back to functioning again and should not be blocked by Cloudflare in the future

**Comment by mlhale7** [comment]
@orangewolf - thanks for this. I'll reply back after I've tested.

**Comment by mlhale7** [comment]
@orangewolf - it does work with the command line utility I referenced using https://samvera:hyku@digitalcollections.lib.utk.edu/catalog/oai as the OAI-PMH endpoint (as suggested by @kirkkwang). Retrieval is much slower than it would be typically. I'm assuming [https://digitalcollections.lib.utk.edu/catalog/oai](https://digitalcollections.lib.utk.edu/catalog/oai )will work once we go live?

Since I mentioned Repox in the description, I want to note that I can't get Repox to resolve the endpoint and show available sets (regardless of which URL I've used). Below are screenshots. Will this resolve when we go live?

![Repox_OAIDoesn'tExist1](https://github.com/user-attachments/assets/f0e4762f-c9ac-440c-b6bd-76edb42a0c14)

![Repox_OAIDoesn'tExist2](https://github.com/user-attachments/assets/73e79151-ecdc-4046-8133-f332743b6180)

**Comment by josh-morgan117** [comment]
@mlhale7 , @kirkkwang , @orangewolf I'm guessing repox is unable to access due to the user login popup. Since you get a 401 for trying to validate against https://www.openarchives.org/Register/ValidateSite

**Comment by mlhale7** [comment]
@kirkkwang just turned off the basic auth. The OAI-PMH URL (https://digitalcollections.lib.utk.edu/catalog/oai) is passing in the validator.

![Image](https://github.com/user-attachments/assets/f613b004-f1ef-493a-b104-37155129a627)

I'm not able to get a OAI-PMH URL to resolve in Repox. Here are screenshots:

![Image](https://github.com/user-attachments/assets/fe0bcc59-3b03-4fe6-966d-814b5ba4110c)

![Image](https://github.com/user-attachments/assets/08ba871a-76e0-40f0-800a-9d0ad6572163)

I did also try removing the "s" from https just in case, but this made no difference.

I'm going to ask Will if there's anything in our Repox logs that might help us troubleshoot this issue. I'm also going to get in touch with the Samvera community through Slack to see if others use Repox and have had similar issues with their OAI-PMH endpoint.

**Comment by mlhale7** [comment]
Testing `https://digitalcollections.lib.utk.edu/catalog/oai` with the authentication off with @kirkkwang today worked! Repox is able to identify the OAI-PMH endpoint and give the existing sets.

<img width="844" alt="Image" src="https://github.com/user-attachments/assets/acf11b32-773b-4bb0-8ef2-6d950aa5f2ef" />
